Block Library: Move `supports` to `block.json` files for core blocks

gziolo opened the pull request:
## Description
I still consider breaking down this PR into smaller tasks because it contains the following refactorings:
- Moves `supports` to `block.json` files for core blocks
- Moves `parents` to `block.json` files for core blocks
- Adds a default variation with initial attributes set to translatable strings to Search block to make the block definition static
- Introduces `block.json` for the following blocks:
  - Navigation
  - Search
- Removes the script that generates `server-registered.json` to use with integration tests and replaces it with the data loaded from `block.json` files

## How has this been tested?
`npm run test-unit`
`npm run test-e2e`

Manually tested that Navigation and Search block works as before.

## Types of changes
Refactoring

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

gziolo committed:
Block Library: Move supports to block.json

gziolo committed:
Blocks: Add "features" alias for "supports" field

gziolo committed:
Block Library: Move "supports" to block.json files

gziolo committed:
Remove server-register.json generator in favor of reading block.json files

Soean commented on docs/rfc/block-registration.md:
```diff
@@ -337,8 +354,8 @@ wp.blocks.registerBlockType( 'my-block/name', {
 	supports: {
-		html: false
-	}
+		html: false,
```
Should we also remove `supports` from this example?

gziolo replied:
Yes, we should. I missed it.

By the way, I accidentaly enabled quite aggressive Visual Studio Code formatting for Markdown files.

ZebulanStanphill reviewed (commented):
I couldn't find any mistakes in the `block.json` files, so those look good to go.

The `// TODO: Decide how to handle it.` comments seem rather vague to me. I think we should make them a bit more descriptive if we're not going to resolve them in this PR.

gziolo commented:
> The `// TODO: Decide how to handle it.` comments seem rather vague to me. I think we should make them a bit more descriptive if we're not going to resolve them in this PR.

I'm going to remove them and open an issue with a list to address it separately. It isn't clear to me why it can't be handled in the place that handles the `supports` flag instead.

gziolo commented:
> > The `// TODO: Decide how to handle it.` comments seem rather vague to me. I think we should make them a bit more descriptive if we're not going to resolve them in this PR.
> 
> I'm going to remove them and open an issue with a list to address it separately. It isn't clear to me why it can't be handled in the place that handles the `supports` flag instead.

Blocks that should be tackled separately because of the condition in `supports`:
https://github.com/WordPress/gutenberg/blob/7452701b8eafb2fce7c28124e5b7645d8bc2f78b/packages/block-library/src/columns/index.js#L30
https://github.com/WordPress/gutenberg/blob/7452701b8eafb2fce7c28124e5b7645d8bc2f78b/packages/block-library/src/group/index.js#L96
https://github.com/WordPress/gutenberg/blob/7452701b8eafb2fce7c28124e5b7645d8bc2f78b/packages/block-library/src/heading/index.js#L38

With that, I can remove those comments :)

gziolo committed:
Remove temporary TODO code comments

gziolo commented:
@Soean and @ZebulanStanphill – thanks for your feedback, I addressed both issues.

gziolo commented:
There is one more thing to clarify since I mentioned that I could split this PR into smaller parts. Given that I had to rebase this PR already, I would be more than happy to land all changes in one batch given how they all are interconnected. In fact, this is mostly "extract and move" refactoring.

youknowriad commented:
About the mobile checks to disable the supports, I think that's probably a wrong approach, instead they should update the hooks to "not apply" if needed or apply differently and not disable the support entirely.

gziolo commented:
> About the mobile checks to disable the supports, I think that's probably a wrong approach, instead they should update the hooks to "not apply" if needed or apply differently and not disable the support entirely.

I'll propose that as an independent PR.

youknowriad commented on packages/block-library/src/search/variations.js:
```diff
@@ -0,0 +1,14 @@
+		isDefault: true,
+		attributes: { buttonText: __( 'Search' ), label: __( 'Search' ) },
+	},
+];
```
That's an interesting usage of "Variations" :)

gziolo replied:
I tried the same approach with File block but it uses placeholders so it loses the attribute that is sourced from HTML after save and refresh because it isn't stored with `save` method. 

There are drawbacks :)

gziolo replied:
A note (mostly to myself) for future reference. We could try to experiment with `supports: { placeholder: { type: 'media' } }`. With that flag, it would show media placeholder and when saving it would serialize attributes sources from HTML as attributes serialized in HTML comment of the block. It would resolve the issue with variations. The same could work with `supports: { placeholder: { type: 'variations' } }` for Columns block but it would show the variations picker when `save` returns null.

oandregal reviewed (commented):
Hi! I see this is already approved but I have a few comments/questions about this that I wanted to share:

- Is it the goal to replace `register_block_type` by `register_block_type_from_metadata`? The first is still used in the `post_comments` block.

- variations for the search block: what if we also added `placeholder` for completeness? I know it's not translated yet so it's not really necessary. It's fine as it is also.

- supports key: besides all the blocks with conditionals you mentioned, there's also the `widget-area` block. Can we migrate this to block.json as well?

- I saw that we added a parent for the `query-loop` and `query-pagination` blocks. It makes sense to me but also changes the behavior so I wanted to confirm this was intentional.

- We also added `html:false` for search and post-tags. Not sure what's the criteria for not allowing editing the html of block so wanted to flag to confirm this is intentional.

It's fine by me if we want to do this in follow-up PRs (if they make sense).

gziolo committed:
Improve registration on the server for the Post Comments block

gziolo committed:
Move supports to block.json for Widget Area block

gziolo commented:
> * Is it the goal to replace `register_block_type` by `register_block_type_from_metadata`? The first is still used in the `post_comments` block.

Yes, ideally we use `register_block_type_from_metadata` (which still uses `register_block_type` internally). New commit b5f607fbcd comming up soon.

> * variations for the search block: what if we also added `placeholder` for completeness? I know it's not translated yet so it's not really necessary. It's fine as it is also.

It's an empty string so I think it's fine to leave it as is.

> * supports key: besides all the blocks with conditionals you mentioned, there's also the `widget-area` block. Can we migrate this to block.json as well?

It was added recently, it's hard to keep up but here you go 9ba3e7598a.

> * I saw that we added a parent for the `query-loop` and `query-pagination` blocks. It makes sense to me but also changes the behavior so I wanted to confirm this was intentional.

It might be a rebase issue, I will double-check. It was in the JS file before.

> * We also added `html:false` for search and post-tags. Not sure what's the criteria for not allowing editing the html of block so wanted to flag to confirm this is intentional.

I think it's expected since those blocks don't define `save` so it's defaulting to `null`. There is nothing to edit in HTML mode :)

> It's fine by me if we want to do this in follow-up PRs (if they make sense).

I think I addressed all those points :)

gziolo committed:
Fix rebase issue for Query blocks

gziolo commented:
I opened https://github.com/WordPress/gutenberg/pull/22502 to refactor the remaining blocks listed in https://github.com/WordPress/gutenberg/pull/22422#issuecomment-630977935.

gziolo commented:
Thank you all for help 💯
